lib/puppet/type/concat_file.rb

```diff
@@ -295,7 +295,11 @@ def nested_merge(hash1, hash2)
 
   def fragment_content(r)
     if r[:content].nil? == false
-      fragment_content = r[:content]
+      fragment_content = if r[:content].is_a?(Puppet::Pops::Types::PSensitiveType::Sensitive)
```

Comment:
I'm a little surprised this is necessary.  I could swear I'm already using `Sensitive` `content` and other than a warning about not being able to automatically redact stuff it's 'working'.  (I still have to manually set `show_diff => false` on the `concat` resource, but no unwrapping anything anywhere.)

Reply:
Actually, I'm not.  I'm setting `content` to a `node_encrypt::secret` which is a function that returns a `Deferred`.  The deferred function decrypts the data and _that_ returns a `Sensitive` though. https://github.com/puppetlabs/puppetlabs-node_encrypt/blob/8b1a02f1847e582b58145a0c145f89771e1975bf/lib/puppet/functions/node_decrypt.rb#L12

So I still get the warning `Unable to mark 'content' as sensitive: content is a parameter and not a property, and cannot be automatically redacted.`, but the lacking of unwrapping in the type doesn't break anything for me.

Reply:
Given the following manifest
```puppet
concat { '/tmp/testconcat':
  show_diff      => false,
  ensure_newline => true,
}

concat::fragment { 'frag1':
  target  => '/tmp/testconcat',
  content => '# a normal header',
  order   => '01',
}
concat::fragment { 'frag2':
  target  => '/tmp/testconcat',
  content => Sensitive('# my password'),
  order   => '02',
}
concat::fragment { 'frag3':
  target  => '/tmp/testconcat',
  content => '# a normal footer',
  order   => '03',
}
```
`r[:content].class` is apparently a `String` in all 3 cases.

Reply:
I'm struggling to replicate the error in spec tests, but the following somewhat convoluted example triggers it.
```puppet
concat { '/tmp/testconcat':
  ensure_newline => true,
}

concat::fragment { 'frag1':
  target  => '/tmp/testconcat',
  content => Deferred('unwrap', ['# some content']),
}
```

```
Error: /Stage[main]/Main/Concat[/tmp/testconcat]/Concat_file[/tmp/testconcat]: Failed to generate additional resources using 'eval_generate': can't modify frozen String
/home/halexfis/concat_test2/concat/lib/puppet/type/concat_file.rb:317:in `fragment_content'
/home/halexfis/concat_test2/concat/lib/puppet/type/concat_file.rb:212:in `block in should_content'
/home/halexfis/concat_test2/concat/lib/puppet/type/concat_file.rb:211:in `each'
/home/halexfis/concat_test2/concat/lib/puppet/type/concat_file.rb:211:in `should_content'
/home/halexfis/concat_test2/concat/lib/puppet/type/concat_file.rb:356:in `eval_generate'
/home/halexfis/.rvm/gems/ruby-2.5.3/gems/puppet-6.29.0/lib/puppet/transaction/additional_resource_generator.rb:56:in `eval_generate'
```

Since with this change to support `sensitive` directly it will probably be much more likely, maybe you want to include a fix for this??

```ruby
fragment_content = fragment_content.dup << newline unless %r{#{newline}\Z}.match?(fragment_content)
```

instead of
```ruby
fragment_content << newline unless %r{#{newline}\Z}.match?(fragment_content)
```
should do it.

Reply:
I've added another commit and attributed to you :)